functions/go/apply-replacements/replacements.go

```diff
@@ -0,0 +1,56 @@
+package main
+
+import (
+	"fmt"
+	"io"
+
+	sdk "github.com/GoogleContainerTools/kpt-functions-catalog/thirdparty/kyaml/fnsdk"
+	"sigs.k8s.io/kustomize/api/filters/replacement"
+	"sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/kustomize/kyaml/fn/framework"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+)
+
+type Replacements struct {
+	Replacements []types.Replacement `json:"replacements,omitempty" yaml:"replacements,omitempty"`
+}
+
+const fnConfigKind = "Replacements"
+
+// Config initializes Replacements from a functionConfig sdk.KubeObject
+func (r *Replacements) Config(functionConfig *sdk.KubeObject) error {
+	if functionConfig.Kind() != fnConfigKind {
+		return fmt.Errorf("received functionConfig of kind %s, only functionConfig of kind %s is supported",
+			functionConfig.Kind(), fnConfigKind)
+	}
+	r.Replacements = []types.Replacement{}
+	if err := functionConfig.As(r); err != nil {
+		return fmt.Errorf("unable to convert functionConfig to %s:\n%w",
+			"replacements", err)
+	}
+	return nil
+}
+
+// Run accepts a reader and writer, and calls framework.Execute to run the function. This exists
+// to enable the function to be invoked as a library.
+func (r *Replacements) Run(reader io.Reader, writer io.Writer) error {
+	rw := &kio.ByteReadWriter{
+		Reader:                reader,
+		Writer:                writer,
+		KeepReaderAnnotations: true,
+	}
+	return framework.Execute(r, rw)
+}
+
+// Process implements framework.ResourceListProcessor interface.
+func (r *Replacements) Process(rl *framework.ResourceList) error {
+	if err := r.Config(sdk.NewFromRNode(rl.FunctionConfig)); err != nil {
+		return err
+	}
+	transformedItems, err := replacement.Filter{Replacements: r.Replacements}.Filter(rl.Items)
+	if err != nil {
+		return err
+	}
+	rl.Items = transformedItems
+	return nil
```

Comment:
Suggest recording the errors in rl.Results field. Here's some examples https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/master/functions/go/set-image/main.go#L25

Reply:
These errors already get recorded in the Results field when run by kpt. Kpt has some logic to look at the returned error and store it in Results.

For example, currently running `kpt fn render` on an error package gives:

```
Package "apply-replacements-error": 
[RUNNING] "gcr.io/kpt-fn/apply-replacements:unstable"
[PASS] "gcr.io/kpt-fn/apply-replacements:unstable" in 1.1s
  Results:
    [error]: nothing selected by Pod.[noVer].[noGrp]/my-po.[noNs]:spec
```

If we record the error in rl.Results in the function logic itself, we will duplicate the error:

```
Package "apply-replacements-error": 
[RUNNING] "gcr.io/kpt-fn/apply-replacements:unstable"
[PASS] "gcr.io/kpt-fn/apply-replacements:unstable" in 1.1s
  Results:
    [error]: nothing selected by Pod.[noVer].[noGrp]/my-po.[noNs]:spec
    [error]: nothing selected by Pod.[noVer].[noGrp]/my-po.[noNs]:spec
```

I suggest we just return the error when we get one and let the orchestrator (kpt) handle it, rather than recording it again.

Reply:
Hmmm, I don't think kpt appends error (sdk does though, but only when the Process function returns err), and the apply-replacements error should results in `[FAIL]` not `[PASS]`. Mind sharing the code about the duplicate errors case?

Reply:
> Hmmm, I don't think kpt appends error (sdk does though, but only when the Process function returns err)

I see. Yeah, since Process function is returning an error, the error is already getting appended to the Results. 

> Mind sharing the code about the duplicate errors case?

```
// Process implements framework.ResourceListProcessor interface.
func (r *Replacements) Process(rl *fn.ResourceList) error {
	if err := r.Config(rl.FunctionConfig); err != nil {
		return err
	}
	transformedItems, err := r.Transform(rl.Items)
	if err != nil {
		rl.Results = append(rl.Results, &fn.Result{
			Message: err.Error(),
			Severity: fn.Error,
		})
		return err
	}
	rl.Items = transformedItems
	return nil
}
```

Changing `Process` to the above (appending the error to Results) results in the error being duplicated in the Results field because the returned error is appended elsewhere after Process returns the error. 

> the apply-replacements error should results in [FAIL] not [PASS]

I am not sure why this is happening, but it is definitely strange. I will investigate.

Reply:
Oh I see, I didn't realize we didn't want to return the error. Made the change, thanks!